### DISCUSSION TITLE
Document Store API Feature Request

### DISCUSSION THREAD

**Discussion by NOGIT007** [FlowiseAI/Flowise#2910] [Ideas]
### Document Store API Feature Request
Objective: Enhance Document Store capabilities to include dynamic data management and integration with Google Drive for seamless document handling, thereby improving chatbot or AI interaction systems' efficiency and adaptability.

### Required API Endpoints:
Add Document: An endpoint to add new documents to the Document Store, allowing users to expand the database dynamically.
Replace Document: An endpoint to replace existing documents within the Document Store. This feature should allow updating the document contents without changing its reference ID or location within the store.
Delete Document: An endpoint to delete documents from the Document Store. This function should remove the document entirely from the database, ensuring that the data remains current and relevant.

### Integration Features:
Google Drive Integration: Integrate the Document Store API with Google Drive, allowing documents stored in Google Drive to be automatically synced with the Document Store. This integration should support real-time updates, so any changes made in Google Drive are automatically reflected in the Document Store.
No UI Development Required: By leveraging Google Drive's existing interface, users can manage documentation without the need for a separate, custom UI. This approach simplifies documentation management, allowing employees to use familiar tools and workflows.

### Use Cases Enhancement:
Dynamic RAG Model Updating: With these new API endpoints, users can easily manage the Document Store's content, enabling dynamic updates to the RAG (Retriever-Augmented Generation) model. This feature is crucial for maintaining an up-to-date and efficient chatbot or virtual agent system.
Clean Data Maintenance: The ability to add, replace, and delete documents directly through the API ensures that the RAG model can always operate with the most current and relevant data. This capability significantly enhances the chatbot's or agent's accuracy and relevance in user interactions.

### Expected Benefits:
Improved Efficiency: Streamlining document management and integration processes enhances overall system efficiency.
Enhanced Flexibility: The proposed features promote system adaptability to different use cases, including chatbots and customer service agents.
User Familiarity: Utilizing Google Drive avoids the need for additional training or adaptation, as it leverages a widely-used platform for document management.

### Conclusion
Implementing these features would significantly enhance the Document Store's functionality, making it a more powerful tool for managing dynamic content in AI-driven applications. By enabling direct document management through API endpoints and integrating with commonly used platforms like Google Drive, the system becomes more user-friendly and adaptable to various use cases.

Of course pimp with AI :)

**Comment by mdwoicke** [comment]
Definitely much needed! 👍

**Comment by ricardogonzalezcases** [comment]
Is there an alternative? I need to synchronize a knowledge program dynamically! 
Thanks!

**Comment by HenryHengZJ** [comment]
Google drive integration is coming up in next release!

**Comment by fmancardi** [comment]
I'm interested in improvement in API to have Create, Update an Delete end points.

**Create/Add** 
Right now exists and works 99.99% (post /document-store/upsert/{id}).
I said 99.99% because after the upsert you are not able to access the Options -> Preview and Process 
(I've open this #5097)  

**Update**
post /document-store/upsert/{id} have a replaceExistent option but according to documentation does not update the Vector Database (no good)

**Delete** 
It will be useful.


Right now only options for update and delete is to call the **refresh whole document store API**

**Reply by m3svcs** [reply]
Dear sirs

I did some API tests.

POST /document-store/upsert/{id} updates the Document Store with the Data Loader, and also updates the Vector Store.

DELETE /document-store/loader/{storeId}/{loaderId} deletes the document via its {storeid} and {loaderid} but does not update the Vector Store.  That is according to the API documentation online.

DELETE /document-store/vectorstore/{id} is supposed to remove (or synchronise the deleted data from {storeid} with the Vector Store, but I am puzzled by the lack of additional data for this operation.  Does this API call DELETE **ALL** data pertaining to the {storeid} from the Vector Store since there could be several other {loaderid} in the Vector Store?

Thank you all.

**Comment by fmancardi** [comment]
@m3svcs 

IMHO there is also clarification regarding  [document store refresh API](https://docs.flowiseai.com/api-reference/document-store#post-document-store-refresh-id)

**post/document-store/refresh/{id}**

Is needed.
<img width="931" height="888" alt="Screenshot 2025-08-28 152707" src="https://github.com/user-attachments/assets/82c9a701-05b2-4acf-8356-32d73b2c82cc" />


Is possible to use the body data to request the refresh of just a set of document loaders?
But the documentation states: 
Re-process and upsert **all existing documents in document store**

In the body you can provide same kind of configuration for the loaders, that is used when you call upsert API end point.
Then at least for me it's not clear to what extent this is similar o different to the upsert API end point.


More detailed documentation and examples are needed in order to be able to use the API effectively

**Reply by m3svcs** [reply]
@fmancardi

Thank you for your kind attention to my query.  Yes, I am aware of post/document-store/refresh/{id}

I have thousands of thousands of document loaders to take care of and I'm detecting duplicates in the chunks and programmatically identify and remove them automatically.  I have assigned a curator for the job.  The person just clicks and search for repeats or bad or irrelevant documents to remove them.  I thought there might be a more efficient way of doing it.

Specify a folder and upsert hundreds of document at a time using:
  POST /document-store/upsert/{id}

Specify a {loaderid} in the (storeidd} to delete 'bad' documents using:
  DELETE /document-store/loader/{storeId}/{loaderId}

As mentioned earlier, AI suggested using the following to remove **a single document** from the Vector Store:
  DELETE /document-store/vectorstore/{id} where {id} is {storeid}

I am afraid **all** the data pertaining to the {storeid} in {vectoreStoreId} might be deleted due to that call.  I only could verify all the data in the storeid but not the vectorStoreId because I only know how to read into Postgres database. not the Faiss.Index file.

All my Postgres validation of data seems fine till date.

Thank you for this great project.  Appreciate it very much.

**Comment by fmancardi** [comment]
@m3svcs 
I would like to share experiences regarding using the API for upsert & refresh.
Do not think if this discussion will be the right place.
Anyway I'm going to describe what is happening in my installation.
Using flowise free on premises, with OpenAI emmbeddings, Qdrant with Record Manager (Postgres).
Here the upsert configuration

<img width="949" height="912" alt="qdrant_setup-blured" src="https://github.com/user-attachments/assets/af5121e6-1b35-4145-ae36-c91f10d41061" />


**I've followed this flowise documentation advice**

<img width="1012" height="148" alt="image" src="https://github.com/user-attachments/assets/d058db00-6659-4880-8da7-38accb4048c9" />

Then I've created a template document uploader, then used it for the upsert REST API call.  
I've done the upsert grouping my documents, using an script to upsert each group.
I'm a novice regarding using Qdrant, then my choice while the upsert procceded, was monitoring using the qdrant dashboard the number of points in the collection.
I see an strange thing, number of points increased but when running the script for another group of documents
**it does not continue to increase** but it seems restarted from 1.

When the whole group scripts finished, I decided to access the document store and launch the **Upsert All Chunks.**

When the upsert all ended, the point number was increased as expected.  
This is very confusing for me.
It would be great is someone can provide an explanation.

**Reply by m3svcs** [reply]
@fmancardi 3.

This is my simple version for upserting.  I too am a novice trained in COBOL of the 70s and 80s:
<img width="1346" height="982" alt="image" src="https://github.com/user-attachments/assets/35f4555a-592a-46d6-bfce-be30dbd7c7fb" />

I followed Leon's Video Guide on his YouTube Channel to install Flowise V3.0.5 for ragging on my friend's home PC.  The completed database will then be served worldwide from his home.

What you described above is indeed puzzling.  It seems to me:
   POST /document-store/refresh/{id}
is the only SAFE way to ensure data integrity for the system.

Thank you once again.

**Comment by m3svcs** [comment]
Hi fmancardi

I managed to do some Vector Store synchronisation testing after deleting loader id from the Document Store, I found the safest but slow way is by going with your suggestion:
    post/document-store/refresh/{id}

Thank you very much.

**Comment by m3svcs** [comment]
Hi fmancardi

An update:
post/document-store/refresh/{id} doesn't always work, whenever it fails with an error, I'll need to re-enter the UI and re-upsert ALL Chunks manually.

Cheers!

**Reply by fmancardi** [reply]
@m3svcs 
Thanks for the info.
IMHO the API feature needs lot of attention from the dev team.
You can look for the issues I've open (filter by fmancardi).
Unfortunately the solutions do not arrive to the rate I would like and trying to contribute making changes to the code is not easy. :(
